tests/components/esphome/test_config_flow.py

```diff
@@ -41,6 +45,159 @@
+    mock_storage.async_get_key = AsyncMock(return_value=stored_key)
+    mock_storage_func.return_value = mock_storage
+
+    # Verify the key can be retrieved (simulating config flow behavior)
+    retrieved_key = await mock_storage.async_get_key(mac_address)
+    assert retrieved_key == stored_key
+
+
+@patch("homeassistant.components.esphome.config_flow.async_get_encryption_key_storage")
+async def test_retrieve_encryption_key_from_storage_no_mac_address(
+    mock_storage_func, hass: HomeAssistant
+) -> None:
+    """Test _retrieve_encryption_key_from_storage with no MAC address."""
+
+    mock_storage = Mock()
+    mock_storage_func.return_value = mock_storage
+
+    flow = EsphomeFlowHandler()
+    flow.hass = hass
+    flow._device_mac = None
+    flow._reauth_entry = None
+
+    # Should return False when no MAC address is available
+    result = await flow._retrieve_encryption_key_from_storage()
+    assert result is False
+
+    # Storage should not be called
+    mock_storage_func.assert_not_called()
+
+
+@patch("homeassistant.components.esphome.config_flow.async_get_encryption_key_storage")
+async def test_retrieve_encryption_key_from_storage_with_device_mac(
+    mock_storage_func, hass: HomeAssistant
+) -> None:
+    """Test _retrieve_encryption_key_from_storage using _device_mac."""
+
+    stored_key = "test_encryption_key_32_bytes_long"
+    mock_storage = Mock()
+    mock_storage.async_get_key = AsyncMock(return_value=stored_key)
+    mock_storage_func.return_value = mock_storage
+
+    flow = EsphomeFlowHandler()
+    flow.hass = hass
+    flow._device_mac = "11:22:33:44:55:aa"
+    flow._reauth_entry = None
+    flow._noise_psk = None
+
+    # Should return True and set noise_psk
+    result = await flow._retrieve_encryption_key_from_storage()
+    assert result is True
+    assert flow._noise_psk == stored_key
+
+    # Storage should be called with the device MAC
+    mock_storage.async_get_key.assert_called_once_with("11:22:33:44:55:aa")
+
+
+@patch("homeassistant.components.esphome.config_flow.async_get_encryption_key_storage")
+async def test_retrieve_encryption_key_from_storage_with_reauth_entry(
+    mock_storage_func, hass: HomeAssistant
+) -> None:
+    """Test _retrieve_encryption_key_from_storage using reauth entry unique_id."""
+
+    stored_key = "test_encryption_key_32_bytes_long"
+    mock_storage = Mock()
+    mock_storage.async_get_key = AsyncMock(return_value=stored_key)
+    mock_storage_func.return_value = mock_storage
+
+    # Create a mock reauth entry
+    mock_entry = Mock()
+    mock_entry.unique_id = "aa:bb:cc:dd:ee:ff"
+
+    flow = EsphomeFlowHandler()
+    flow.hass = hass
+    flow._device_mac = None  # No device MAC
+    flow._reauth_entry = mock_entry
+    flow._noise_psk = None
+
+    # Should return True and set noise_psk using reauth entry
+    result = await flow._retrieve_encryption_key_from_storage()
+    assert result is True
+    assert flow._noise_psk == stored_key
```

Comment:
Its generally frowned upon to test the flow handler directly.  Instead we should use the public api to setup the config flow and step through it